README
 
             Enterprise Class Event Logging for Linux
                   ========================================
Content
--------
 1. Introduction
 2. How to install
 3. Known Bugs
 4. License
 5. Version
 6. Change History
 7. Reporting Bugs
 8. Authors



1. Introduction
---------------

A full description of Enterprise-Class Event logging can be found at:

  http://evlog.sourceforge.net/

Not all of the features described in the specification have been
provided.  For a list of these features, see the following:

  http://evlog.sourceforge.net/not_yet_available.html
  
	
2. How to install
------------------

See the INSTALL file.



3. Known Bugs
-------------

See bug tracker on the project page for more details.

530006   Forwarding of syslog msgs = xdm failure
         (This problem has not been observed on Red Hat 8.0
          nor SuSE SLES8 (UL 1.0)
586100   ppc64: evlview fails with long, pointer


4. License 
----------

The kernel patch is licensed under the GNU General Public License (GPL)
Version 2, June 1991.  The full text of the GPL is located at:

evlog/kernel/GPL

The evlogd daemon is also licensed under the GPL.

---

The user library is licensed under the GNU Lesser Public License (LGPL)
Version 2.1, February 1999.  The full text of the LGPL is located at:

evlog/user/LGPL


5. Latest Version
-----------------

1.6.0



6. Change History
-----------------

See the CHANGES file.



7. Reporting Bugs
-----------------
If you encounter a bug, please submit it to the project bug list,
located on the project page:

   https://sourceforge.net/projects/evlog/



8. Authors
-----------

Jim Keniston    kenistoj@users.sourceforge.net
Larry Kessler   lkessler@users.sourceforge.net
Hien Nguyen     nguyhien@users.sourceforge.net



CHANGES
Enterprise-Class Event Logging for Linux
========================================

v1.0.0   9/20/2001
Initial  Version


v1.0.1  10/01/2001
- Replaced calls to spin_lock_irq() and spin_unlock_irq() with calls
  to  spin_lock_irqsave() and spin_unlock_irqrestore() in kernel log
  write functions.  This resolved a hang condition experienced with 
  a SCSI device driver.
- Replaced use of sockaddr to sockaddr_un to provide longer socket 
  path name.
- Several minor changes needed for s390 (32 and 64 bit).
- runtests.sh scripts modifed to take option of s390 (32-bit) or
  s390x (64-bit) to specify s390/IBM Z-Series.
- Processing of white-spaces was fixed to correctly process >1 white 
  space.
NOTE:  The patch for GNU C Library, evlog-1.0.1_glibc2.2.3.patch,
is unchanged from version 1.0.0.


v1.1.0  11/09/2001

New features and enhancements in v1.1.0
    
1. Kernel patches for kernel versions 2.4.2 through 2.4.14 will 
   now be posted on the website.  
2. A new library that "wraps" around GNU C Lib and provides
   forwarding of syslog messages to the POSIX event log without
   having to patch and rebuild glibc.
3. Formatting Template support and evltc command (for compiling
   formatting template). Some features are not yet available
   (section numbers in the spec are indicated):
* (6.1.2.2.1.2: Imports and Typedefs) If an import statement specifies
   a struct that has already been created or imported, the import
   statement will be silently ignored.
* (6.1.2.2.1.4: Attributes Section [overall format]) The issue regarding
   const-only structs is still unresolved.  It is currently impossible 
   to define a struct that has only const attributes.
* (6.1.2.2.1.4.3: Attribute Dimension or Bit-Field Width) Bit fields 
   and arrays of structs are not yet supported.
* (6.1.2.2.1.4.5 Attribute Format)   The following is not yet supported:
* Validation of printf-style formats is incomplete.  In particular, you
  are permitted to specify formats that are completely inappropriate for
  the given data type -- e.g. "%f" for a string.
* The %t (dump) format is currently supported only for arrays.
* %b (bitmap) and %v (named-value) specifiers are not yet supported
  for arrays.
* As previously mentioned, arrays of structs are not yet supported;
  so the %Z format specifier is not, either.
* (6.2.3.3 Formatting Templates [API]) The following is not supported:
   * The evl_populatetemplate() function does not currently create the 
     _EXTRA_DATA_ pseudo-attribute
   * The evltemplate_getatt() function does not currently handle names 
     such as "x.y" to reference members of struct-type attributes.

4. Displaying of events logged in the kernel (with the new log write
   functions) on the system console.  The --output option was added to 
   the evlconfig command to turn console logging on/off, or set the
   severity level for displaying to the console.
5. Forwarding of printk() messages to the kernel event buffer must now
   be explicitly enabled when configuring the kernel (forwarding is
   disabled by default).
6. evlfacility command for managing the facility registry.
7. Restricted Logging support. 
8. Added "address" as a data type for formatting templates.
9. Added 2 new purpose arguments for posix_log_query_create() as
   extensions to the draft POSIX standard:  EVL_PRPS_TEMPLATE and
   EVL_PRPS_RESTRICTED.
10. Added EVL_KERNEL_EVENT flag for log_flags member in 
    posix_log_entry.
11. Added varargs versions of log write functions in kernel and 
    user-space.
12. Added support for --new and --timeout options in evlview
    command.
13. Added --templates and --notemplates options to evlview command.
14. Added unlink(PidFile) function in evlogd.c, evlnotifyd.c,
    evlactiond.c.
15. Moved evlogd, evlnotifyd, and evlactiond pid files from /var/evlog
    to /var/run.

Bug Fixes in v1.1.0

1. Removed evlog *.pid files on graceful shutdown or on startup if 
   previous shutdown was non-graceful to fix evlogd daemon startup
   failure.
2. In event notification, fixed a bug where error returns from
   rt_sigqueueinfo() were causing the notification to be erroneously
   disabled.
3. In do_syslog(), now use options 20 and 21 for read/write to kernel
   event buffer, instead of 9 and 10, to avoid conflict with other
   newly added options in do_syslog(). 
4. posix_log_notify_add() is now signal-safe.


v1.1.1  11/28/2001

Bug fixes in v1.1.1

1. (Fix for Bug # 485676) Code added in v1.1.0 for displaying events on the
   system console is incompatible with the re-write of code for displaying of
   messages from printk() first appearing in kernel v2.4.10. This causes the
   kernel to hang or PANIC under certain conditions.  This only affects the
   patches for kernel versions 2.4.10 through 2.4.14. 
2. Added check in the kernel version of posix_log_write() for null-terminated
   string (when log_format is POSIX_LOG_STRING).  Returns EBADMSG if string
   is not null-terminated.  A few other changes to make this function   
   consistent between kernel and user space.
3. Fixed bug causing programs linked with libevl.so to fail at link time.
4. Added missing function call in libevlsyslog.so so that syslog messages
   actually get logged in event log.
5. Fixed several bugs causing problems on s390 and/or 64-bit machines:
   * evltc command reporting bogus errors or causing segfaults
   * Backward seeks through the event log (as used by evlview --tail) failed. 
   * Various memory corruption problems 
6. Updated all header files for C++ compatability.


v1.2.0  02/18/2002

New Features and Enhancements in v1.2.0

1. Events logged in user-space are now passed directly to the evlogd 
   daemon,  instead of being written to the kernel event buffer.
2. The libevl event logging library no longer depends on the event 
   logging kernel patch, meaning that event logging in user-space 	
   does not require  changes in the kernel.  The kernel patch is 

user/include/evlog.h
/*
 * Linux Event Logging for the Enterprise
 * Copyright (c) International Business Machines Corp., 2001
 *
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 *  Please send e-mail to lkessler@users.sourceforge.net if you have
 *  questions or comments.
 *
 *  Project Website:  http://evlog.sourceforge.net/
 *
 */

#ifndef _EVLOG_H_
#define _EVLOG_H_

#include <sys/types.h>
#include <stdarg.h>
#include <string.h>

#include <linux/evl_log.h>

#ifdef __cplusplus
extern "C" {
#endif
#ifndef EVLOG_STATE_DIR
#define EVLOG_STATE_DIR			"/var/evlog"
#endif
#define EVLOGD_SERVER_SOCKET		EVLOG_STATE_DIR "/evlogdsocket"
#define EVLOGD_EVTSERVER_SOCKET		EVLOG_STATE_DIR "/evlogdevtsocket"
#define EVLNOTIFYD_SERVER_SOCKET	EVLOG_STATE_DIR "/evlnotifydsocket"
#define EVLOG_CONF_SOCKET		EVLOG_STATE_DIR "/evlconfsoc"
#define EVLACTIOND_SERVER_SOCKET	EVLOG_STATE_DIR "/evlactiondsocket"

/* Record formats.  See also <linux/evl_log.h> */
#ifndef POSIX_LOG_PRINTF
#define POSIX_LOG_PRINTF	3
#endif

/* Output formats */
#define EVL_COMPACT 0x1

typedef int evlrecsize_t;
#define SIZESIZE sizeof(evlrecsize_t)

/*
 * Log Management Event Types
 * These are event types that may be logged for facility LOG_LOGMGMT.
 *
 * The following event types are reserved, and are defined in other headers:
 * 1-5: These are the standard LOG_LOGMGMT event types defined in POSIX 1003.25
 * -- e.g., POSIX_LOG_MGMT_STARTMAINT.
 * 40: EVLOG_REGISTER_FAC -- A request from the kernel to register a facility.
 */

/* Event types 1-5 are reserved. */

#define EVLOG_FORK_FAILED       8
#define EVLOG_WRITE_PID         9
#define EVLOG_PID_EXISTS        10
#define EVLOG_SIG_ACT           11
#define EVLOG_OPEN_LOG_FAILED   12
#define EVLOG_WRITE_LOG_FAILED  13
#define EVLOG_READ_LOG_FAILED   14
#define EVLOG_SEEK_LOG_FAILED   15
#define EVLOG_MALLOC_FAILED     16
#define EVLOG_READBUF_FAILED    17

#define EVLOG_WRITE_REQF_FAILED		20
#define EVLOG_READ_REQF_FAILED		21
#define EVLOG_OPEN_REQF_FAILED		22
#define EVLOG_OPEN_SOCKET_FAILED	23
#define EVLOG_BROKEN_PIPE			24
#define EVLOG_UID_OP_FAILED			25
#define EVLOG_GID_OP_FAILED			26
#define EVLOG_FAILED_APPENDTOLIST	27

#define EVLOG_PWFILE_LOOKUP_FAILED 	28
#define EVLOG_GRFILE_LOOKUP_FAILED 	29

#define EVLOG_LOADREG_FAILED		30
#define EVLOG_ACCESS_DENIED			31

/* Event type 40 is reserved. */

#define NFY_ACCESS_DENIED		0xfa
#define NFY_ACCESS_GRANTED		0xac
#define NFY_MAX_CLIENTS			0xca

#define EVLOGD_ACCESS_DENIED		NFY_ACCESS_DENIED
#define EVLOGD_ACCESS_GRANTED		NFY_ACCESS_GRANTED
#define EVLOGD_MAX_CLIENTS			NFY_MAX_CLIENTS

/* 
 * Architecture signature for each log record.
 * This will be the lower word of the of the rec->log_magic field.
 */
#define LOGREC_NO_ARCH				0x0000
#define LOGREC_ARCH_I386			0x0001
#define	LOGREC_ARCH_IA64			0x0002
#define LOGREC_ARCH_S390			0x0003
#define LOGREC_ARCH_S390X			0x0004
#define LOGREC_ARCH_PPC				0x0005
#define LOGREC_ARCH_PPC64			0x0006
#define LOGREC_ARCH_X86_64			0x0007
#define LOGREC_ARCH_ARM_BE			0x0008
#define LOGREC_ARCH_ARM_LE			0x0009

/*
 * Logfile Header structure.
 */
typedef struct log_header {
	uint		log_magic;	/* Magic number indicating log file */
	long		log_version;	/* Event logging version */
	posix_log_recid_t last_recId;	/* Last recid in log: used by evlogd */
	off_t		reserved1;
	uint		log_generation;	/* Changes during log maintenance */
} log_header_t;

/* log an event */
extern int evl_log_write(posix_log_facility_t facility, int event_type,
	posix_log_severity_t severity, unsigned int flags, ...);

extern int evl_log_vwrite(posix_log_facility_t facility, int event_type,
	posix_log_severity_t severity, unsigned int flags, va_list args);

extern int evl_pack_format_and_args(const char *format, va_list args,
	char *databuf, size_t *datasz);

/* format fixed portion of event record */
extern int evl_format_evrec_fixed(const struct posix_log_entry *entry,
	char *buf, size_t buflen, size_t *reqlen, const char *separator,
	size_t linelen, int fmt_flags);

/* format variable portion of event record */
extern int evl_format_evrec_variable(const struct posix_log_entry *entry,
	const void *var_buf, char *buf, size_t buflen, size_t *reqlen);

/* format event record according to user-supplied format */
extern int evl_format_evrec_sprintf(const struct posix_log_entry *entry,
	const void *var_buf, const char *format, char *buf,
	size_t buflen, size_t *reqlen);

/* compute facility code from name */
extern int evl_gen_facility_code(const char *fname,
	posix_log_facility_t *fcode);

/* add facility to registry, if needed, and get code */
extern int evl_register_facility(const char *fname,
	posix_log_facility_t *fcode);

/* For internal use... */
extern int _evlDumpBytes(const void *data, size_t nBytes, char *buf,
	size_t buflen, size_t *reqlen);
extern int _evlDumpBytesForce(const void *data, size_t nBytes, char *buf,
	size_t buflen, size_t *reqlen);
extern size_t _evlGetMaxDumpLen();
extern int _evlFormatPrintfRec(const char *data, size_t datasz, char *buf,
	size_t buflen, size_t *reqlen, int printk);

/* The syslog equivalent of the kernel's printkat() */
extern int _evl_syslogat(int priority, const char *facname, int event_type,
	const char *fmt, const char *unbraced_fmt, ...);
extern char * _evl_unbrace(const char *fmt);
extern int evl_gen_event_type(const char *s1, const char *s2, const char *s3);
#ifndef __Lynx__
extern void syslog(int priority, const char *fmt, ...);
#else
extern void syslog(int priority, char *fmt, ...);
#endif

#define __stringify2(s) #s
#define __stringify(s) __stringify2(s)

struct evlog_position {
   int line;
   char function[64 - sizeof(int)];
   char file[128];
};

#define _EVLOG_POS { __LINE__, __FILE__ }

struct evlog_info {
   char format[128+64];
   char facility[64];
   struct evlog_position pos;
};

#define syslogat(priority, fmt, ...) \
    do { \
	static struct evlog_info __attribute__((section(".log"),unused)) ___ \
		= { fmt, __stringify(EVL_FACILITY_NAME), _EVLOG_POS }; \
	char *unbraced_fmt = _evl_unbrace(fmt); \
	if (_evl_syslogat(priority, __stringify(EVL_FACILITY_NAME), \
	    evl_gen_event_type(__FILE__, __FUNCTION__, fmt), \
	    fmt, unbraced_fmt, ##__VA_ARGS__) != 1) \
		syslog(priority, unbraced_fmt, ##__VA_ARGS__); \
    	free(unbraced_fmt); \
    } while (0)

#ifdef __cplusplus
}
#endif

#endif  	/* _EVLOG_H_ */
